fix (revised) for issue with CardDAV sync with the DAVDroid sync client.

ethoms opened the pull request:
This is a revised fix for the issue raiased in sogo bug tracker 3370 and 3373. It supercedes the fix in commit  2c723070c69899055b456119f6e8f42c279d0c57 .

The fix was noted in NEWS with the comment "we now return all cards when we receive an empty addressbook-query REPORT". However it did not work for me and at least two others, as can be seen in the commit comments. In summary, only contacts with email addresses were synced. The suggested change from kwirk fixes the regular address book sync, but it completely breaks syncing of the read-only Group Directory (Corporate Directory). My suggested changes work in full (as far as I'm able to test).

I have done some fairly extensive testing of CardDAV sync (with DAVDroid only) and it seems to work 100% now. In addition to the obvious tests, I have tested with contacts that only have one field of data entered. The feilds I've tested (with all other fields empty) are as follows:

First name
Last name
Display name
email address
Work (telephone)
Home (telephone)
Fax (telephone)
Mobile (telephone)

Additionally, I tested syncing of a contact with only the Work Address fully populated. In the webmail, since the name fields are all missing, the "Organization" field of the Work Address takes the place of the name field in the 'Name' column. This does get synced to my phone and it also appears my Android contact list with 'Name' set to the 'Organization' field data. The address, organization and website fields being in tact also.

In addition, I tested a Group Directory (Corporate Directory) [SOGoUserSources->isAddressBook] sync. It seems contacts without email addresses do not sync. This seems to be the behaviour across the board with a "." search filter. This happens despite the filter in SOGoUserSources including ldap entries without a mail attribute. Nothing I can do to patch this in SOGoFolder+CardDAV.m, that would have to be fixed in the code that deals with the special "." search filter (I guess).

I think the contact search system needs some looking into, particularly the "." search filter behaviour. There is another bug related to contact search in the webmail address book view. I will make a bug report on that soon. It's a shame there isn't an "all" search filter, it would seem it would make various parts of SOGo easier to get the right behaviour.

kwirk commented:
@ethoms - May be worth double checking, but for me this patch seems to cause duplicate entries for contacts with email addresses :smile:

ethoms commented:
Hmm, not exactly. Almost all my contacts are perectly fine. But I do have one that has a duplication issues. It's very strange. I can't see any pattern why that contacts is different. It has duplicate fields but only one contact. But like I said, the vast majority are all OK.

ethoms commented:
Update, that one contact is OK now. I deleted the contact and added it back fresh, no duplicate now. Perhaps it's a knowck-on effect from a previous sync when the sync went wrong.

ethoms commented:
Are they duplicated in the webmail too? Or just the Android device? If just the Android device, then try to do the following for the DAVDroid app: 1.) clear cache, 2.) clear data 3.) uninstall 4.) install from F-Droid if possible, otherwise wherever you normally install from. Then try again.

ethoms commented:
So, I have found more contacts that have duplicate telephone numbers. I delete the duplicate, sync, and it seems OK. Once I get done checking them all and deduplicating, I will uninstall DAVDroid, reinstall and see if a fresh sync behaves well.

kwirk commented:
I was using _curl_ to see the direct response back from sogo, as per [sogo issue 3370](http://sogo.nu/bugs/view.php?id=3370). I wasnt sure what havoc this may cause with DAVDroid...

ethoms commented:
OK, so are you saying that when you simulate a sync with *curl*, the results show duplicates? Could you give the full command (substituting sensitive data) so I can try it my end.

Perhaps this is not a problem in a real sync as the sync app ignores duplicates. In any case my DVdroid is syncing fine after I deleted a few duplicated fields. To me it looked like the same field data was coming both ways. Once I deleted one of the duplicate fields, it does not come back and everything is fine.

After deleteing my DAVdroid accounts and reinstalling DAVdroid, my first fresh sync was perfectly fine, no duplicates at all. So it's possible that if I had just done this step alone, I may not have needed to remove duplicate fields by hand.

So, DAVdroid at least does not mind if there are duplicate contacts in the sync results from the server. I will have to test other CardDAV tools to confirm.

ethoms commented:
Just tested a fresh sync with another CardDAV sync tool called "CardDAV Sync free (by Marten Gajda)". It works flawlessly too. So my patch is alright by my tests.

ethoms commented:
Double check you applied the patch exactly. One filter is using a "." the other is using a "%".

kwirk commented:
@ethoms -  I was using the _curl_ command as per [sogo issue 3370](http://sogo.nu/bugs/view.php?id=3370). Patch was downloaded direct from this PR.

It's worth noting that SOGo already sends contacts twice in two slightly different versions for compatibility reasons (DAVDroid only processes one), so with this PR you actually get 4 copies. As well as bandwidth considerations, I'd be concerned about how different applications would handle duplicates.

I think the fact that SOGo doesn't de-duplicate results (ignoring the different versions behaviour) when using multiple filters is a bug in its own right which will need to be fixed first.
